packages/core/src/DatadogProvider/Buffer/BoundedBuffer.ts

```diff
@@ -0,0 +1,112 @@
+import { DatadogBuffer } from './DatadogBuffer';
+
+const DEFAULT_BUFFER_SIZE = 100;
+
+type VoidCallback = {
+    callback: () => Promise<void>;
+    _type: 'VOID';
+};
+
+type CallbackReturningId = {
+    callback: () => Promise<string>;
+    returnedBufferId: string;
+    _type: 'RETURNING_ID';
+};
+
+type CallbackWithId = {
+    callback: (id: string) => Promise<void>;
+    withBufferId: string;
+    _type: 'WITH_ID';
+};
+
+type BufferItem = VoidCallback | CallbackReturningId | CallbackWithId;
+
+export class BoundedBuffer extends DatadogBuffer {
+    private bufferSize: number;
+    private buffer: BufferItem[] = [];
+    private idTable: { [bufferId: string]: string | null } = {};
+
+    constructor(bufferSize: number = DEFAULT_BUFFER_SIZE) {
+        super();
+        this.bufferSize = bufferSize;
+    }
+
+    addCallback = (callback: () => Promise<void>) => {
+        if (this.buffer.length < this.bufferSize) {
+            this.buffer.push({ callback, _type: 'VOID' });
+        }
+
+        return new Promise<void>(resolve => resolve(undefined));
+    };
+
+    addCallbackReturningId = (callback: () => Promise<string>) => {
+        const bufferId = this.generateRandomBufferId();
+        if (this.buffer.length < this.bufferSize) {
+            this.buffer.push({
+                callback,
+                returnedBufferId: bufferId,
+                _type: 'RETURNING_ID'
+            });
+            this.idTable[bufferId] = null;
+        }
+
+        return new Promise<string>(resolve => resolve(bufferId));
+    };
+
+    addCallbackWithId = (
+        callback: (id: string) => Promise<void>,
+        bufferId: string
+    ) => {
+        if (this.idTable[bufferId] !== undefined) {
+            this.buffer.push({
+                callback,
+                withBufferId: bufferId,
+                _type: 'WITH_ID'
+            });
+        }
+
+        return new Promise<void>(resolve => resolve(undefined));
+    };
+
+    drain = async () => {
+        for (
+            let bufferIndex = 0;
+            bufferIndex < this.buffer.length;
+            bufferIndex++
+        ) {
+            const item = this.buffer[bufferIndex];
+            if (item._type === 'RETURNING_ID') {
```

Comment:
Does it make sense to declare `VOID`, `RETURNING_ID` and `WITH_ID` as constant strings somewhere, or as a TypeScript enum?

Reply:
With type safety and autocomplete, it's much easier to use union types than enums, with the same added value. (for more in-depth comparison: https://www.bam.tech/article/should-you-use-enums-or-union-types-in-typescript)

The only case where I keep Enums is for anything that is customer-facing, as they could work without typescript.